addon/transforms/enum.js

```diff
@@ -1,22 +1,62 @@
-import Transform from 'ember-data/transform';
-import Enum from 'ember-enum/enum';
 import Ember from 'ember';
+import Transform from 'ember-data/transform';
+import { ERROR_MESSAGES } from 'ember-enum/enum';
+import createEnumType from 'ember-enum/utils/create-enum-type';
+
+const {
+  assert,
+  computed,
+  get,
+  isArray,
+  warn
+} = Ember;
 
-const { typeOf } = Ember;
+const ENUM_OPTIONS_MUST_BE_DEFINED = `ENUM ERROR: when using the enum data type,
+you must define the 'options' array in the options object passed to
+ember-data's 'attr' function.`;
+
+const SERVER_RETURNED_INVALID_VALUE = `ENUM WARN: Server returned an invalid
+value for the enum. ${ERROR_MESSAGES.INVALID_VALUE}`;
+
+function warnIncorrectValue(options) {
+  let errorMsg = SERVER_RETURNED_INVALID_VALUE;
+  errorMsg += options.join(', ');
+  warn(errorMsg, false, {
+    id: 'Ember-Enum.warn-incorrect-server-value',
+    url: 'https://github.com/rmachielse/ember-enum'
+  });
+}
 
 export default Transform.extend({
-  deserialize(serialized, { options = [], defaultValue = null } = {}) {
-    return Enum.create({
-      value: serialized || defaultValue,
-      options
+  deserialize(serialized, { options, defaultValue }) {
+
+    assert(ENUM_OPTIONS_MUST_BE_DEFINED, isArray(options));
+    let EnumType   = createEnumType(serialized, options, defaultValue);
+    let enumObject = EnumType.create();
+    let value      = serialized;
+
+    if (!enumObject.isValidOption(value)) {
+      warnIncorrectValue(serialized, options);
+      value = defaultValue || options[0];
+    }
+
+    enumObject.set('value', value);
+
+    return computed({
+      get() {
+        return enumObject;
+      },
+
+      set(key, value) {
+        this.propertyWillChange(key);
+        enumObject.set('value', value);
+        this.propertyDidChange(key);
+        return enumObject;
+      }
     });
   },
 
   serialize(deserialized) {
-    if (typeOf(deserialized) === 'instance') {
-      return deserialized.get('value');
-    } else {
-      return deserialized;
-    }
+    return get(deserialized, 'value') || deserialized.toString() || deserialized;
```

Comment:
Don't think `deserialized` could be undefined since mutation just sets the value. *Maybe* if the user calls `delete`? seems like a huge anti-pattern though.

ah, good catch. we probably don't want `toString()` since we want that to be the handlebars rendering hook IMO.